Use SIRI-ET and GTFS-RT TripUpdates at the same time

leonardehrenfried opened the pull request:
### Summary

Previously it was not possible to run SIRI-ET and GTFS-RT TripUpdates in the same instance. This PR changes it by cleaning up the relationships of the various real-time components.

A class diagram to show the new relationships is available in the issue #6348.

### Issue

Closes #6348

### Unit tests

Some unit tests updated.

### Documentation

The new class diagram is available in this PR in `doc/dev/diagrams/TimetableSnapshotManager.svg`.

### Bumping the serialization version id

While no actual entities have changed, but the `TimetableRepository` has which is serialized into the graph. Therefore we need to bump it.

cc @fahrplaner

vpaturet reviewed (changes requested):
This looks good. Just minor comments.

leonardehrenfried commented on application/src/main/java/org/opentripplanner/transit/service/DefaultTransitService.java:
```diff
@@ -533,9 +536,6 @@ public boolean hasNewTripPatternsForModifiedTrips() {
    */
   @Nullable
   private TimetableSnapshot lazyGetTimeTableSnapShot() {
```
It absolutely could. Can you live with the messy diff if I do that?

vpaturet replied:
I can live with that diff :-)

leonardehrenfried commented on application/src/main/java/org/opentripplanner/updater/spi/TimetableSnapshotFlush.java:
```diff
@@ -14,28 +15,17 @@ public class TimetableSnapshotFlush implements Runnable {
-      if (gtfsTimetableSnapshotSource != null) {
-        gtfsTimetableSnapshotSource.flushBuffer();
-      }
-      LOG.debug("Flushed timetable snapshot buffer");
```
I'll fix that.

vpaturet commented on application/src/main/java/org/opentripplanner/transit/service/DefaultTransitService.java:
```diff
@@ -183,9 +186,8 @@ public RegularStop getRegularStop(FeedScopedId id) {
-    TimetableSnapshot currentSnapshot = lazyGetTimeTableSnapShot();
-    if (currentSnapshot != null) {
-      Route realtimeAddedRoute = currentSnapshot.getRealtimeAddedRoute(id);
+    if (this.timetableSnapshot != null) {
```
Not very important, but why do you prefix the field with "this"?

leonardehrenfried replied:
I didn't do anything - it was all IntelliJ! :)

leonardehrenfried replied:
https://github.com/opentripplanner/OpenTripPlanner/pull/6363/commits/c740b8302d914d2a6c7bb0f59284018ddfad6631

vpaturet reviewed (approved):
You probably don't need to bump the serialization id since you change only a transient field.

leonardehrenfried commented:
> You probably don't need to bump the serialization id since you change only a transient field.

The tag is already set.

leonardehrenfried commented:
So removing fields doesn't count as breaking backwards-compatibility?

leonardehrenfried commented:
@Jouca This comment has nothing to do with the pull request here. Please open a separate issue if you want to request a feature.

abyrd commented:
> So removing fields doesn't count as breaking backwards-compatibility?

I think the idea is that the transient field is not serialized, so the serialized graph is still compatible and will load without error.

leonardehrenfried commented:
Alright then. The label is gone.

abyrd commented on application/src/main/java/org/opentripplanner/updater/siri/SiriTimetableSnapshotSource.java:
```diff
@@ -45,12 +42,12 @@
  * a specific point in time.
  */
-public class SiriTimetableSnapshotSource implements TimetableSnapshotProvider {
+public class SiriTimetableSnapshotSource {
```
Am I correct in understanding that the [Gtfs]TimetableSnapshotSource and SiriTimetableSnapshotSource are no longer sources of timetable snapshots at all, but have become translating or mapping code which interprets messages in two different incoming formats, applying both of them to the same modification API on a single instance of TimetableSnapshotManager injected by Dagger, particularly via `snapshotManager.updateBuffer(realTimeTripUpdate)`?

leonardehrenfried replied:
Yes, that is absolutely correct.

However, since this PR is complex enough as it is, I chose to defer any kind of renaming to a follow up where I will undoubtedly produce a lot of discussion.

abyrd reviewed (commented):
Looking good so far. Just added a few questions that would reassure me I'm understanding what I'm seeing. In addition to those line comments, would it be correct to say that the refactoring described in the Javadoc on TimetableSnapshotManager has now been carried out within this PR? Should that Javadoc be updated?

leonardehrenfried commented on application/src/main/java/org/opentripplanner/transit/service/DefaultTransitService.java:
```diff
@@ -80,20 +80,24 @@ public class DefaultTransitService implements TransitEditorService {
    */
-  private TimetableSnapshot timetableSnapshot;
+  @Nullable
+  private final TimetableSnapshot timetableSnapshot;
```
This is an artifact of the previous way of doing things which looped the timetable snapshot through the `TimetableRepository` and stored a refrerence lazily. But even then the method name referred to was incorrect and should have been `lazyGetTimeTableSnapShot`.

leonardehrenfried commented:
> Looking good so far. Just added a few questions that would reassure me I'm understanding what I'm seeing. In addition to those line comments, would it be correct to say that the refactoring described in the Javadoc on TimetableSnapshotManager has now been carried out within this PR? Should that Javadoc be updated?

These observations are absolutely correct and I have updated the Javadoc.

abyrd reviewed (approved):
Thanks for confirming those points. This looks like a step in the right direction to me, thanks for all your work on this. Understood about any renaming / documentation updates coming in a subsequent step, which I'd be happy to contribute to.
